[AMDGPU] Remove implicit definition of register group when restoring the last sub-register after a spill

bababuck opened the pull request:
Extension of https://reviews.llvm.org/D141101 to remove the define flag from the last stack memory access. Fixes case where COPY instructions are used for some of the stack restoration, but the copies get optimized away during the machine-cp pass.

Prior to this change, was possible to produce the following code:
  $agpr16_agpr17_agpr18_agpr19 = SCRATCH_LOAD_DWORDX4_ST 64, 0, implicit $exec, implicit $flat_scr, implicit-def $agpr16_agpr17_agpr18_agpr19_agpr20_agpr21_agpr22_agpr23_agpr24_agpr25_agpr26_agpr27_agpr28_agpr29_agpr30_agpr31 :: (load (s128) from %stack.17, align 4, addrspace 5)
  $agpr20_agpr21_agpr22_agpr23 = SCRATCH_LOAD_DWORDX4_ST 80, 0, implicit $exec, implicit $flat_scr :: (load (s128) from %stack.17 + 16, align 4, addrspace 5)
  $agpr24_agpr25_agpr26_agpr27 = SCRATCH_LOAD_DWORDX4_ST 96, 0, implicit $exec, implicit $flat_scr :: (load (s128) from %stack.17 + 32, align 4, addrspace 5)
  $agpr31 = COPY $agpr112, implicit $agpr16_agpr17_agpr18_agpr19_agpr20_agpr21_agpr22_agpr23_agpr24_agpr25_agpr26_agpr27_agpr28_agpr29_agpr30_agpr31
  $agpr30 = COPY $agpr208, implicit $agpr16_agpr17_agpr18_agpr19_agpr20_agpr21_agpr22_agpr23_agpr24_agpr25_agpr26_agpr27_agpr28_agpr29_agpr30_agpr31
  $agpr28_agpr29 = SCRATCH_LOAD_DWORDX2_ST 112, 0, implicit $exec, implicit $flat_scr, implicit-def $agpr16_agpr17_agpr18_agpr19_agpr20_agpr21_agpr22_agpr23_agpr24_agpr25_agpr26_agpr27_agpr28_agpr29_agpr30_agpr31 :: (load (s64) from %stack.17 + 48, align 4, addrspace 5)

where `$agpr30 = COPY $agpr208` would be optimized away by `machine-cp` pass. Instead, change to:
  $agpr28_agpr29 = SCRATCH_LOAD_DWORDX2_ST 112, 0, implicit $exec, implicit $flat_scr :: (load (s64) from %stack.17 + 48, align 4, addrspace 5)

Fixes #131386.

Made the simple fix, but I'm not completely comfortable with this change since the reason for the previous inclusion of `IsLastSubReg` is unclear to me.

@krzysz00

bababuck committed:
[AMDGPU] Remove implicit definition of register group when restoring the last sub-register after a spill

Extension of https://reviews.llvm.org/D141101 to remove the define flag from
the last stack memory access. Fixes case where COPY instructions are used
for some of the stack restoration, but the copies get optimized away during the
machine-cp pass.

Prior to this change, was possible to produce the following code:
  $agpr16_agpr17_agpr18_agpr19 = SCRATCH_LOAD_DWORDX4_ST 64, 0, implicit $exec, implicit $flat_scr, implicit-def $agpr16_agpr17_agpr18_agpr19_agpr20_agpr21_agpr22_agpr23_agpr24_agpr25_agpr26_agpr27_agpr28_agpr29_agpr30_agpr31 :: (load (s128) from %stack.17, align 4, addrspace 5)
  $agpr20_agpr21_agpr22_agpr23 = SCRATCH_LOAD_DWORDX4_ST 80, 0, implicit $exec, implicit $flat_scr :: (load (s128) from %stack.17 + 16, align 4, addrspace 5)
  $agpr24_agpr25_agpr26_agpr27 = SCRATCH_LOAD_DWORDX4_ST 96, 0, implicit $exec, implicit $flat_scr :: (load (s128) from %stack.17 + 32, align 4, addrspace 5)
  $agpr31 = COPY $agpr112, implicit $agpr16_agpr17_agpr18_agpr19_agpr20_agpr21_agpr22_agpr23_agpr24_agpr25_agpr26_agpr27_agpr28_agpr29_agpr30_agpr31
  $agpr30 = COPY $agpr208, implicit $agpr16_agpr17_agpr18_agpr19_agpr20_agpr21_agpr22_agpr23_agpr24_agpr25_agpr26_agpr27_agpr28_agpr29_agpr30_agpr31
  $agpr28_agpr29 = SCRATCH_LOAD_DWORDX2_ST 112, 0, implicit $exec, implicit $flat_scr, implicit-def $agpr16_agpr17_agpr18_agpr19_agpr20_agpr21_agpr22_agpr23_agpr24_agpr25_agpr26_agpr27_agpr28_agpr29_agpr30_agpr31 :: (load (s64) from %stack.17 + 48, align 4, addrspace 5)

where `$agpr30 = COPY $agpr208` would be optimized away by `machine-cp` pass.
Instead, change to:
  $agpr28_agpr29 = SCRATCH_LOAD_DWORDX2_ST 112, 0, implicit $exec, implicit $flat_scr :: (load (s64) from %stack.17 + 48, align 4, addrspace 5)

Fixes #131386.

bababuck commented on llvm/test/CodeGen/AMDGPU/accvgpr-spill-scc-clobber.mir:
```diff
@@ -4111,10 +4111,10 @@ body:             |
   ; GFX908-NEXT:   BUFFER_STORE_DWORD_OFFSET killed $vgpr1, $sgpr0_sgpr1_sgpr2_sgpr3, $sgpr32, 0, 0, 0, implicit $exec :: (store (s32) into %stack.2, addrspace 5)
   ; GFX908-NEXT:   $vgpr1 = V_MOV_B32_e32 8200, implicit $exec
-  ; GFX908-NEXT:   $vgpr0 = V_ACCVGPR_READ_B32_e64 $agpr0, implicit $exec, implicit-def $agpr0_agpr1
+  ; GFX908-NEXT:   $vgpr0 = V_ACCVGPR_READ_B32_e64 $agpr0, implicit $exec, implicit-def $agpr0_agpr1, implicit $agpr0_agpr1
```
There seems to be a side effect where more implicit register reads are now included, unclear to me why this occurs. I will dig further to try to understand unless someone here has any thoughts.

bababuck commented:
If there is any interest, here's the emulator I wrote to help with debug: https://github.com/bababuck/AMDGPU

bababuck commented:
Ping

Adding @arsenm, apologies is this is out of your area, but you seem to be involved in the AMDGPU backend.

krzysz00 reviewed (commented):
This looks like a straightforward bugfix to me but I don't understand the backend well enough to approve it

krzysz00 commented:
Pinging the people Github suggested

arsenm reviewed (approved):
Reduced version of the failing situation with the lowered spill + machine-cp would be good

bababuck committed:
[AMDGPU] New MIR test with register spillage/restoration but some sub-reg's are saved to agpr's

Replicates #131386.

Currently, this test "fails", note the missing copy to $agpr14.

bababuck commented:
Updated with the following changes:
- Added new MIR test that reduces the failing scenario into a few lines of MIR code
- Update the source code change to only change the if statement (no formatting changes to the surrounding code)
- Rather than dropping `IsLastSubReg`, I instead qualify it with checking that the register state is not a define. I noticed in the new test I wrote that an `implicit-kill` was being dropped after my original change.

arsenm reviewed (commented):
First line description should only contain direct description, not the issue number

arsenm commented on llvm/test/CodeGen/AMDGPU/agpr-spill-copy.mir:
```diff
@@ -0,0 +1,43 @@
+--- |
+  define amdgpu_kernel void @agpr_spill_copy() #0 { ret void }
+
+  attributes #0 = { "amdgpu-num-vgpr"="32" }
```
Avoid using amdgpu-num-vgpr. Use amdgpu-waves-per-eu (does this even need a budget restriction? There's no actual allocation)

bababuck replied:
Without a budget, the registers will spill from `vgpr` into available `agpr` registers rather than onto the stack. The bug only occurs when there is enough register pressure such that some of the registers in the register group spill onto the stack, and some spill into the `agprs`. Technically, I could remove the budget restriction, but that would require adding code to the test to put all of the `agprs` into use (that seems much messier to me).

I made the change to using `amdgpu-waves-per-eu` in this latest push, just curious as to why this is preferred? It seems like a more indirect way of getting the behavior that I want (which is limiting the number of `vgpr`'s to `32`.

arsenm replied:
amdgpu-num-vgpr is hacky and introduces a lot of edge cases that don't actually work, and I want to remove it. It's not future proof when the number of registers change

bababuck replied:
Isn't using `amdgpu-waves-per-eu` also not robust to if the number of registers change in the context of this test? Because I'm relying on the fact that there are 256 registers so with `amdgpu-waves-per-eu=8`, 32 registers are allocated for each wave, which is the condition I need for the failing case to occur. Is there a better way to write the test to avoid this dependency; my end goal is to have exactly 2 unused `agpr`'s when the register spill occurs.

arsenm commented on llvm/test/CodeGen/AMDGPU/agpr-spill-copy.mir:
```diff
@@ -0,0 +1,43 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 5
+# RUN: llc -mtriple=amdgcn--amdhsa -mcpu=gfx942 -verify-machineinstrs -run-pass prologepilog,machine-cp -o - %s | FileCheck -check-prefix=GFX942 %s
```
```suggestion
# RUN: llc -mtriple=amdgcn--amdhsa -mcpu=gfx942 -run-pass prologepilog,machine-cp -o - %s | FileCheck -check-prefix=GFX942 %s
```

bababuck replied:
This is updated with the latest push.

arsenm commented:
> First line description should only contain direct description, not the issue number

Still should fix description

arsenm commented on llvm/test/CodeGen/AMDGPU/agpr-spill-copy.mir:
```diff
@@ -0,0 +1,43 @@
+--- |
+  define amdgpu_kernel void @agpr_spill_copy() #0 { ret void }
+
+  attributes #0 = { "amdgpu-waves-per-eu"="8" }
```
```suggestion
  attributes #0 = { "amdgpu-waves-per-eu"="8,8" }
```

bababuck commented:
Fixed:
- Issue and commit titles to no longer include issue number in the first line
- `"amdgpu-waves-per-eu"="8" ` --> "amdgpu-waves-per-eu"="8,8" 
- Rebased to latest `main`

arsenm commented on llvm/lib/Target/AMDGPU/SIRegisterInfo.cpp:
```diff
@@ -1928,7 +1928,9 @@ void SIRegisterInfo::buildSpillLoadStore(
-    if (NeedSuperRegImpOperand && (IsFirstSubReg || IsLastSubReg))
+    bool IsSrcDstDef = SrcDstRegState & RegState::Define;
+    if (NeedSuperRegImpOperand &&
+        (IsFirstSubReg || (IsLastSubReg && !IsSrcDstDef)))
```
Not sure what the second condition is doing here, why isn't this just IsFirstSubReg && isDef?

bababuck replied:
I believe the IsLastSubReg is needed for other flags, namely `RegState::Kill`. Removing it, the following change is seen in the test I wrote:
```
-    ; GFX942-NEXT: SCRATCH_STORE_DWORDX2_SADDR killed $vgpr12_vgpr13, $sgpr32, 48, 0, implicit $exec, implicit $flat_scr, implicit killed $vgpr0_vgpr1_vgpr2_vgpr3_vgpr4_vgpr5_vgpr6_vgpr7_vgpr8_vgpr9_vgpr10_vgpr11_vgpr12_vgpr13_vgpr14_vgpr15 :: (store (s64) into %stack.0 + 48, align 4, addrspace 5)
+    ; GFX942-NEXT: SCRATCH_STORE_DWORDX2_SADDR killed $vgpr12_vgpr13, $sgpr32, 48, 0, implicit $exec, implicit $flat_scr :: (store (s64) into %stack.0 + 48, align 4, addrspace 5)
```
This seems to be the analog of `implicit def` on the first sub-register.
